PostgreSQL 17 support

Hi,

1.0.1 fails to build against PostgreSQL 17. Log is below. v17beta2 is already out. Can you please take a look? 

Thanks, Devrim


> /usr/bin/clang -Wno-ignored-attributes -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-unused-command-line-argument -Wno-compound-token-split-by-macro -Wno-format-truncation -O2  -I /usr/pgsql-17/include -I. -I./ -I/usr/pgsql-17/include/server -I/usr/pgsql-17/include/internal  -D_GNU_SOURCE -I/usr/include/libxml2 -DWITH_GZFILEOP  -I/usr/include -flto=thin -emit-llvm -c -o pg_failover_slots.bc pg_failover_slots.c
pg_failover_slots.c:103:7: error: redefinition of 'standby_slot_names' with a different type: 'List *' (aka 'struct List *') vs 'char *'
  103 | List *standby_slot_names = NIL;
      |       ^
/usr/pgsql-17/include/server/replication/slot.h:232:26: note: previous declaration is here
  232 | extern PGDLLIMPORT char *standby_slot_names;
      |                          ^
pg_failover_slots.c:238:13: warning: incompatible pointer types passing 'char *' to parameter of type 'List *' (aka 'struct List *') [-Wincompatible-pointer-types]
  238 |                 list_free(standby_slot_names);
      |                           ^~~~~~~~~~~~~~~~~~
/usr/pgsql-17/include/server/nodes/pg_list.h:672:29: note: passing argument to parameter 'list' here
  672 | extern void list_free(List *list);
      |                             ^
pg_failover_slots.c:249:10: warning: incompatible pointer types passing 'char **' to parameter of type 'List **' (aka 'struct List **') [-Wincompatible-pointer-types]
  249 |                                                                  &standby_slot_names);
      |                                                                  ^~~~~~~~~~~~~~~~~~~
/usr/pgsql-17/include/server/utils/varlena.h:31:18: note: passing argument to parameter 'namelist' here
   31 |                                                                   List **namelist);
      |                                                                          ^
pg_failover_slots.c:768:32: error: too few arguments to function call, expected 6, have 4
  767 |                 ReplicationSlotCreate(remote_slot->name, true, RS_EPHEMERAL,
      |                 ~~~~~~~~~~~~~~~~~~~~~
  768 |                                                           remote_slot->two_phase);
      |                                                                                 ^
/usr/pgsql-17/include/server/replication/slot.h:239:13: note: 'ReplicationSlotCreate' declared here
  239 | extern void ReplicationSlotCreate(const char *name, bool db_specific,
      |             ^                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  240 |                                                                   ReplicationSlotPersistency persistency,
      |                                                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  241 |                                                                   bool two_phase, bool failover,
      |                                                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  242 |                                                                   bool synced);
      |                                                                   ~~~~~~~~~~~
pg_failover_slots.c:1137:25: warning: comparison of distinct pointer types ('char *' and 'List *' (aka 'struct List *')) [-Wcompare-distinct-pointer-types]
 1137 |         if (standby_slot_names != cached_standby_slot_names)
      |             ~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~
pg_failover_slots.c:1141:24: warning: incompatible pointer types passing 'char *' to parameter of type 'List *' (aka 'struct List *') [-Wincompatible-pointer-types]
 1141 |                         if (list_member_str(standby_slot_names,
      |                                             ^~~~~~~~~~~~~~~~~~
pg_failover_slots.c:1119:23: note: passing argument to parameter 'l' here
 1119 | list_member_str(List *l, const char *str)
      |                       ^
pg_failover_slots.c:1151:29: warning: incompatible pointer types assigning to 'List *' (aka 'struct List *') from 'char *' [-Wincompatible-pointer-types]
 1151 |                 cached_standby_slot_names = standby_slot_names;
      |                                           ^ ~~~~~~~~~~~~~~~~~~
pg_failover_slots.c:1161:15: warning: incompatible pointer types passing 'char *' to parameter of type 'const List *' (aka 'const struct List *') [-Wincompatible-pointer-types]
 1161 |                 list_length(standby_slot_names) == 0)
      |                             ^~~~~~~~~~~~~~~~~~
/usr/pgsql-17/include/server/nodes/pg_list.h:152:25: note: passing argument to parameter 'l' here
  152 | list_length(const List *l)
      |                         ^
pg_failover_slots.c:1198:39: warning: incompatible pointer types passing 'char *' to parameter of type 'const List *' (aka 'const struct List *') [-Wincompatible-pointer-types]
 1198 |                         wait_slots_remaining = list_length(standby_slot_names);
      |                                                            ^~~~~~~~~~~~~~~~~~
/usr/pgsql-17/include/server/nodes/pg_list.h:152:25: note: passing argument to parameter 'l' here
  152 | list_length(const List *l)
      |                         ^
pg_failover_slots.c:1208:25: warning: incompatible pointer types passing 'char *' to parameter of type 'const List *' (aka 'const struct List *') [-Wincompatible-pointer-types]
 1208 |                                                                            list_length(standby_slot_names));
      |                                                                                        ^~~~~~~~~~~~~~~~~~
/usr/pgsql-17/include/server/c.h:1004:28: note: expanded from macro 'Min'
 1004 | #define Min(x, y)               ((x) < (y) ? (x) : (y))
      |                                         ^
/usr/pgsql-17/include/server/nodes/pg_list.h:152:25: note: passing argument to parameter 'l' here
  152 | list_length(const List *l)
      |                         ^
pg_failover_slots.c:1208:25: warning: incompatible pointer types passing 'char *' to parameter of type 'const List *' (aka 'const struct List *') [-Wincompatible-pointer-types]
 1208 |                                                                            list_length(standby_slot_names));
      |                                                                                        ^~~~~~~~~~~~~~~~~~
/usr/pgsql-17/include/server/c.h:1004:40: note: expanded from macro 'Min'
 1004 | #define Min(x, y)               ((x) < (y) ? (x) : (y))
      |                                                     ^
/usr/pgsql-17/include/server/nodes/pg_list.h:152:25: note: passing argument to parameter 'l' here
  152 | list_length(const List *l)
      |                         ^
pg_failover_slots.c:1212:43: warning: incompatible pointer types passing 'char *' to parameter of type 'const List *' (aka 'const struct List *') [-Wincompatible-pointer-types]
 1212 |                            wait_slots_remaining <= list_length(standby_slot_names));
      |                                                                ^~~~~~~~~~~~~~~~~~
/usr/pgsql-17/include/server/c.h:875:9: note: expanded from macro 'Assert'
  875 |                 if (!(condition)) \
      |                       ^~~~~~~~~
/usr/pgsql-17/include/server/nodes/pg_list.h:152:25: note: passing argument to parameter 'l' here
  152 | list_length(const List *l)
      |                         ^
pg_failover_slots.c:1222:25: warning: incompatible pointer types passing 'char *' to parameter of type 'List *' (aka 'struct List *') [-Wincompatible-pointer-types]
 1222 |                         if (!list_member_str(standby_slot_names, NameStr(s->data.name)))
      |                                              ^~~~~~~~~~~~~~~~~~
pg_failover_slots.c:1119:23: note: passing argument to parameter 'l' here
 1119 | list_member_str(List *l, const char *str)
      |                       ^
11 warnings and 2 errors generated.
make[1]: *** [/usr/pgsql-17/lib/pgxs/src/makefiles/../../src/Makefile.global:1085: pg_failover_slots.bc] Error 1
make[1]: *** Waiting for unfinished jobs....
pg_failover_slots.c:103:7: error: conflicting types for ‘standby_slot_names’; have ‘List *’
  103 | List *standby_slot_names = NIL;
      |       ^~~~~~~~~~~~~~~~~~
In file included from /usr/pgsql-17/include/server/replication/logical.h:15,
                 from /usr/pgsql-17/include/server/replication/decode.h:14,
                 from pg_failover_slots.c:38:
/usr/pgsql-17/include/server/replication/slot.h:232:26: note: previous declaration of ‘standby_slot_names’ with type ‘char *’
  232 | extern PGDLLIMPORT char *standby_slot_names;
      |                          ^~~~~~~~~~~~~~~~~~
pg_failover_slots.c: In function ‘synchronize_one_slot’:
pg_failover_slots.c:767:17: error: too few arguments to function ‘ReplicationSlotCreate’
  767 |                 ReplicationSlotCreate(remote_slot->name, true, RS_EPHEMERAL,
      |                 ^~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-17/include/server/replication/slot.h:239:13: note: declared here
  239 | extern void ReplicationSlotCreate(const char *name, bool db_specific,
      |             ^~~~~~~~~~~~~~~~~~~~~
make[1]: *** [<builtin>: pg_failover_slots.o] Error 1

--- a/pg_failover_slots.c
+++ b/pg_failover_slots.c
@@ -763,7 +763,10 @@ synchronize_one_slot(RemoteSlot *remote_slot)
 		 * We have to create the slot to reserve its name and resources, but
 		 * don't want it to persist if we fail.
 		 */
-#if PG_VERSION_NUM >= 140000
+#if PG_VERSION_NUM >= 170000
+		ReplicationSlotCreate(remote_slot->name, true, RS_EPHEMERAL,
+							  remote_slot->two_phase, false, false);
+#elif PG_VERSION_NUM >= 140000
 		ReplicationSlotCreate(remote_slot->name, true, RS_EPHEMERAL,
 							  remote_slot->two_phase);
 #else

